### ISSUE TITLE
Incorrect remaining time is returned when nanosleep error code is set to EINTR

### ISSUE THREAD

**Issue by anjalirx-intel** [oscarlab/graphene#2260]
## Testfile
nanosleep02

## Description of the problem
This testcase verifies that nanosleep() will be successful to suspend the execution of a process, returns after the receipt of a signal and writes the remaining sleep time into the structure.
In this testcase the process sleeps for 1sec and then it get terminated by kill signal. The remaining time is incorrectly returned in Graphene OS

## Expected results
root@ubuntu-test-base:  ./nanosleep02
TINFO: Timeout per run is 0h 05m 00s
TPASS: nanosleep() returned -1, set errno to EINTR
TPASS: **nanosleep() slept for 1000092us, remaining time difference 80us**

## Actual results
root@ubuntu-test-base: ./pal_loader nanosleep02
TINFO: Timeout per run is 0h 05m 00s
TPASS: nanosleep() returned -1, set errno to EINTR
TFAIL: **nanosleep() remaining time 6996850us, expected 3999759us, diff 2997091us**
TBROK: Test haven't reported results!

**Comment by anjalirx-intel** [comment]
@boryspoplawski The issue is fixed for Graphene Native but Graphene-SGX still fails

/home/intel/jenkins/workspace/local_ci_gsgx_apps_test/graphene/LibOS/shim/test/ltp/ltp_src/lib/tst_test.c:1250: TINFO: Timeout per run is 0h 05m 00s
/home/intel/jenkins/workspace/local_ci_gsgx_apps_test/graphene/LibOS/shim/test/ltp/ltp_src/testcases/kernel/syscalls/nanosleep/nanosleep02.c:57: TPASS: nanosleep() returned -1, set errno to EINTR
/home/intel/jenkins/workspace/local_ci_gsgx_apps_test/graphene/LibOS/shim/test/ltp/ltp_src/testcases/kernel/syscalls/nanosleep/nanosleep02.c:68: TFAIL: nanosleep() remaining time 5000009us, expected 3997585us, diff 1002424us
/home/intel/jenkins/workspace/local_ci_gsgx_apps_test/graphene/LibOS/shim/test/ltp/ltp_src/lib/tst_test.c:1067: TBROK: Test haven't reported results!

Summary:
passed   0
failed   0
skipped  0
warnings 0

**Comment by boryspoplawski** [comment]
The original issue was fixed, the 2nd one is most likely unrelated and tracked in #2479

**Comment by aneessahib** [comment]
this issue is different than 2479. 

what happens here is that, when the child process is asleep (read FUTEX_WAIT), a PID KILL (SIGINT) is sent from the parent. Graphene handles this with a thread wake (read FUTEX WAKE), and a SIGCONT being sent to the child process. 
The FUTEX WAKE succeeds, but the SIGCONT is sent even before the original FUTEX WAIT could gracefully exit so that it could subsequently fill up the remainder sleep duration structure properly. Once SIGCONT is hit, the async 
signal handler adjusts the instruction pointer to return the OCALL with -INTR, and this results with incorrect remainder sleep duration being sent back to the application. 

A few possible solutions
1) return from thread wake only after ensuring the futex wait has returned. OR
2) Dont handle SIGCONT - as in go with the OS default behaviour. I have tried this and it works perfectly with all the related LTP tests now passing properly. But not handling this signal the way it is done today could have other implications that im not sure of.  OR
3) Fix the SIGCONT signal handler to handle this case properly. This might be a larger effort. 

I suggest we either reopen this bug, or file another issue that states "Syscalls wakes up too soon .."

**Comment by dimakuv** [comment]
@boryspoplawski to the rescue. I think that Borys's new PR https://github.com/oscarlab/graphene/pull/2602 solves (or partially solves) the root cause reported by @aneessahib. But I'm not sure, so Borys can comment better.

**Comment by boryspoplawski** [comment]
If I understood the root cause correctly then yes, #2602 should fix it. @aneessahib could you try that PR?

**Comment by aneessahib** [comment]
@boryspoplawski  - #2602 does not fix it.

**Comment by boryspoplawski** [comment]
I was not able to reproduce the issue on #2602 in 60 tries, where as it triggeres on current master in most (~90%) tries.

**Comment by aneessahib** [comment]
ok i see what i missed. I tried clock_nanosleep02 - which failed (should be another issue completely). But nanosleep02 passes now with the PR. @anjalirx-intel could try this on her side once the PR is merged.
